dali/pipeline/operators/expressions/arithmetic.h

```diff
@@ -0,0 +1,280 @@
+  }
+  if (is_left_scalar) {
+    return right;
+  }
+  if (is_right_scalar) {
+    return left;
+  }
+  using std::to_string;
+  DALI_ENFORCE(left == right, "Input shapes of element-wise arithemtic operator \"" + op +
+                                  "\" do not match. Expected equal shapes, got: " + op + "(" +
+                                  to_string(left) + ", " + to_string(right) + ").");
+  return left;
+}
+
+template <typename Backend>
+DLL_PUBLIC kernels::TensorListShape<> PropagateShapes(ExprNode &expr,
+                                                      const workspace_t<Backend> &ws) {
+  if (expr.GetNodeType() == NodeType::Constant) {
+    expr.SetShape(kernels::TensorListShape<>{{1}});
+    return expr.GetShape();
+  }
+  if (expr.GetNodeType() == NodeType::Tensor) {
+    auto &e = dynamic_cast<ExprTensor &>(expr);
+    expr.SetShape(ws.template InputRef<Backend>(e.GetInputIndex()).shape());
+    return expr.GetShape();
+  }
+  if (expr.GetSubexpressionCount() == 2) {
+    auto &func = dynamic_cast<ExprFunc&>(expr);
+    auto left_shape = PropagateShapes<Backend>(func[0], ws);
+    auto right_shape = PropagateShapes<Backend>(func[1], ws);
+    func.SetShape(ShapePromotion(func.GetFuncName(), left_shape, right_shape));
+    return func.GetShape();
+  }
+  DALI_FAIL(make_string("Only binary expressions are supported. Got expression with",
+                        expr.GetSubexpressionCount(), "subexpressions."));
+}
+
+/**
+ * @brief Arithmetic operator capable of executing expression tree of element-wise
+ *        arithmetic operations.
+ *
+ * Only expressions consisting of one function node with tensor inputs are now supported.
+ *
+ * There are 3 levels for unit of work.
+ * - Thread (CPUBackend) or CUDA kernel invokation (GPUBackend)
+ * - Task - group of tiles to process by thread or CUDA kernel
+ * - Tile - describes a portion of linear buffer, we try to split the amount of work
+ *          evenly into tasks.
+ *
+ * For CPUBackend we have fixed number of threads that get to process a number of tasks,
+ * so the work is evenly distributed. For GPUBackend we pack all tiles into 1 task, to limit
+ * the number of CUDA calls.
+ */
+template <typename Backend>
+class ArithmeticGenericOp : public Operator<Backend> {
+ public:
+  inline explicit ArithmeticGenericOp(const OpSpec &spec) : Operator<Backend>(spec) {
+    expr_ = ParseExpressionString(spec.GetArgument<std::string>("expression_desc"));
+  }
+
+ protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    output_desc.resize(1);
+
+    if (!types_layout_inferenced_) {
+      result_type_id_ = PropagateTypes<Backend>(*expr_, ws);
+      result_layout_ = GetCommonLayout<Backend>(*expr_, ws);
+      types_layout_inferenced_ = true;
+    }
+
+    result_shape_ = PropagateShapes<Backend>(*expr_, ws);
+    AllocateIntermediateNodes();
```

Comment:
I guess `AllocateIntermediateNodes` can be done only once for all, no need to redo at every `SetupImpl`.